build_test/lib/src/resolve_source.dart

```diff
@@ -97,32 +115,61 @@ Future<T> resolveSource<T>(
 /// **NOTE**: All `package` dependencies are resolved using [PackageAssetReader]
 /// - by default, [PackageAssetReader.currentIsolate]. A custom [resolver] may
 /// be provided to map files not visible to the current package's runtime.
-Future<T> resolveAsset<T>(AssetId input, FutureOr<T> action(Resolver resolver),
-        {PackageResolver resolver, Future tearDown}) =>
-    _resolveAsset(input, action, resolver: resolver, tearDown: tearDown);
-
-/// Internal only backing implementation of `resolveAsset` and `resolveSource`.
-///
-/// If [inputContents] is non-null, it is used instead of reading [input] from
-/// the file system.
-Future<T> _resolveAsset<T>(AssetId input, FutureOr<T> action(Resolver resolver),
-    {String inputContents, PackageResolver resolver, Future tearDown}) async {
-  resolver ??= PackageResolver.current;
-  var syncResolver = await resolver.asSync;
-  var reader = new PackageAssetReader(syncResolver, input.package);
-  var builder = new _ResolveSourceBuilder(action, tearDown);
-  var inputs = [input];
-  var inMemory = new InMemoryAssetReader(
-    sourceAssets: {
-      input: inputContents ?? await reader.readAsString(input),
+Future<T> resolveAsset<T>(
+  AssetId inputId,
+  FutureOr<T> action(Resolver resolver), {
+  PackageResolver resolver,
+  Future<Null> tearDown,
+}) {
+  return _resolveAssets(
+    {
+      '${inputId.package}|${inputId.path}': _useAssetReader,
     },
-    rootPackage: input.package,
+    inputId.package,
+    action,
+    resolver: resolver,
+    tearDown: tearDown,
+  );
+}
+
+/// Internal-only backing implementation of `resolve{Asset|Source(s)}`.
+///
+/// If the value of an entry of [inputs] is [_useAssetReader] then the value is
+/// instead read from the file system, otherwise the provided text is used as
+/// the contents of the asset.
+Future<T> _resolveAssets<T>(
+  Map<String, String> inputs,
+  String rootPackage,
+  FutureOr<T> action(Resolver resolver), {
+  PackageResolver resolver,
+  Future<Null> tearDown,
+}) async {
+  final syncResolver = await (resolver ?? PackageResolver.current).asSync;
+  final assetReader = new PackageAssetReader(syncResolver, rootPackage);
+  final resolveBuilder = new _ResolveSourceBuilder(action, tearDown);
+  final inputAssets = <AssetId, String>{};
+  await Future.forEach(inputs.keys, (String rawAssetId) async {
```

Comment:
Is there a reason we need to do this serially instead of allowing multiple file reads to start?

Reply:
Nope! Changed to `Future.wait`!